[Feature] Added filtering functionality to UI for annots

andreysaf opened the pull request:
You can now filter out annotations through UI:

![filter](https://user-images.githubusercontent.com/6531389/89471991-4985d400-d734-11ea-9ec7-912aca378324.gif)

ZhijieZhang commented on src/components/FilterAnnotMondal/FilterAnnotModal.js:
```diff
@@ -0,0 +1,152 @@
+    let annotTypesToBeAdded = new Set();
+    annots.forEach(annot => {
+      authorsToBeAdded.add(core.getDisplayAuthor(annot));
+      annotTypesToBeAdded.add(annot.Subject.toLowerCase().replace(/\s/g, ''));
```
I found that Link annotations don't have `Subject`, so opening the filter modal when a document has link annotations will blow up the app.

andreysaf commented on src/components/NotesPanel/NotesPanel.js:
```diff
@@ -361,6 +318,12 @@ const NotesPanel = () => {
+                    dataElement="filterAnnotationButton"
+                    className="filter-annotation-button"
+                    label={t('component.filter')}
+                    onClick={() => dispatch(actions.openElement('filterModal'))}
```
Correct, since the rest is just formatting.

mparizeau-pdftron commented:
@ZhijieZhang For your comment 2 I think it would probably be difficult to always match the custom function to an author or type because the function could in theory have anything in side of it. As long as we have a data-element for the filter button they could hide that filter and use their own custom filter?

andreysaf commented:
@ZhijieZhang 

- unchecking filters now shows previously filtered annots
- blank authors or subjects do not show up
- check for translation and if not use the original subject
- modal can be dismissed by esc key
- modal now focus traps the user into making selections
- added comments and resolved all nit

khein-pdftron commented:
> I didn't review the CSS part, not sure if @khein-pdftron will have some comments on them?
> 
> Other than the comments below, I have two concerns that I think should be addressed.
> 
> 1. Check the box and click the apply button filters the notes in the panel. This is working as expected. But later if I uncheck the same box, and click the apply button, the previous filtered notes don't come back in the panel.
>    I think in this case previous filtered nodes should come back since the box is unchecked.
>    It looks like the current way is to click the `cancel` button is to reset it, but I feel like users also expect that unchecking the box should work.
> 2. `setCustomNoteFilter` is an external API that can be used by our clients. If a client calls `instance.setCustomNoteFilter(annot => annot.Subject !== 'abc')` then annotations with `abc` Subject will not show in the notes panel. But if I open the filter modal, `abc` is not checked. I think in this case `abc` should be checked because it is filtered by the API call.

Thanks @ZhijieZhang. I left Andrey a review for the styling on Xodo. I will approve when those changes are done.

liamross commented on src/components/FilterAnnotMondal/FilterAnnotModal.js:
```diff
@@ -0,0 +1,172 @@
+
+  return isDisabled ? null : (
+    <Swipeable onSwipedUp={closeModal} onSwipedDown={closeModal} preventDefaultTouchmoveEvent>
+      <FocusTrap locked={isOpen}>
```
Looks fantastic! Only suggestion would be to re-focus the "Filter" button when the modal closes (there's a prop for that)

```suggestion
      <FocusTrap locked={isOpen} focusLastOnUnlock>
```

liamross replied:
It does look like that only fixes it if you use escape to close though, let me look into that quickly

ZhijieZhang commented on src/components/FilterAnnotMondal/FilterAnnotModal.js:
```diff
@@ -0,0 +1,172 @@
+          type = typesFilter.includes(annot.Subject.toLowerCase().replace(/\s/g, ''));
+        }
+        if (authorFilter.length > 0) {
+          author = authorFilter.includes(annot.Author);
```
Hmm it looks like this one hasn't been addressed.
```
Should we use the display author for authorFilter.includes? I see that authors contains display authors.
```

ZhijieZhang commented on src/components/FilterAnnotMondal/FilterAnnotModal.js:
```diff
@@ -0,0 +1,172 @@
+        authorsToBeAdded.add(core.getDisplayAuthor(annot));
+      }
+      if (annot.Subject && annot.Subject !== '') {
+        // we want to lowercase it and remove spaces to get i18n translations
```
Thanks for putting the comment. Now I'm wondering if we need to display the translated text?

For annotations created inside WebViewer, their `Subject` have already been translated using i18n. See the code at `setupI18n.js`.
```js
window.Annotations.Utilities.setAnnotationSubjectHandler(type =>
  t(`annotation.${type}`),
);
```
Since it's already translated I feel like it's not necessary to do a `lowerCase` then `replace` here and then render `t(annotation.${val})`. `annotation.Subject` here is already `t(annotation.${val})`.

For annotations created by users in other PDF viewers, we don't have existing translations for their `Subject`s, so we don't need to do these string manipulations.

ZhijieZhang commented on src/components/FilterAnnotMondal/FilterAnnotModal.js:
```diff
@@ -0,0 +1,172 @@
+                              t(`annotation.${val}`)
+                                .toLowerCase()
+                                .replace(/\s/g, '') === val
+                                ? t(`annotation.${val}`)
```
See my another comment regarding if we can just display `annot.Subject` here instead of getting the `val`s and use `t(annotation.${val})`

andreysaf commented:
![image](https://user-images.githubusercontent.com/6531389/90040838-4e3f1080-dc7d-11ea-8a47-55417be25e2a.png)

Fixed CSS, fixed translations, removed closing of the modal on clear, fixed floating of the sorting container, fixed include author vs displayAuthor.

Thanks @ZhijieZhang for taking your time to review!
